File: android/app/src/main/java/org/getlantern/lantern/activity/WebViewActivity.java
package org.getlantern.lantern.activity;

import android.app.ProgressDialog;
import android.view.KeyEvent;
import android.view.View;
import android.webkit.WebSettings.PluginState;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.fragment.app.FragmentActivity;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.Extra;
import org.androidannotations.annotations.ViewById;
import org.getlantern.lantern.LanternApp;
import org.getlantern.lantern.R;
import org.getlantern.lantern.model.ProxySettings;
import org.getlantern.mobilesdk.Logger;

@EActivity(R.layout.webview)
public class WebViewActivity extends BaseFragmentActivity {

    private static final String TAG = WebViewActivity.class.getName();

    private ProgressDialog dialog;

    @Extra
    String url;

    @ViewById
    WebView webView;

    public void closeWebView(View view) {
        this.finish();
    }

    @AfterViews
    void afterViews() {
        if (url == null) {
            Logger.error(TAG, "Not opening webview; invalid url");
            return;
        }
        dialog = new ProgressDialog(this);
        openWebview(url);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_DOWN) {
            switch (keyCode) {
                case KeyEvent.KEYCODE_BACK:
                    if (webView.canGoBack()) {
                        webView.goBack();
                    } else {
                        finish();
                    }
                    return true;
            }
        }
        return super.onKeyDown(keyCode, event);
    }

    protected void showProgressDialog() {
        if (dialog != null) {
            dialog.show();
        }
    }

    protected void hideProgressDialog() {
        if (dialog != null && dialog.isShowing()) {
            dialog.dismiss();
        }
    }


    protected void openWebview(final String url) {
        final String proxyAddr = LanternApp.getSession().getHTTPAddr();
        final String[] parts = proxyAddr.split(":");
        if (parts.length > 0) {
            final String proxyHost = parts[0];
            final int proxyPort = Integer.parseInt(parts[1]);
            ProxySettings.setProxy(this,
                    webView, proxyHost, proxyPort);
        }
        Logger.debug(TAG, "Opening " + url + " in webview");

        // Added to get around an issue where we might be unable
        // to start an intent without a user gesture. This allows
        // the webview to redirect automatically to the payment page
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                return false;
            }
        });

        webView.getSettings().setLoadWithOverviewMode(true);
        webView.getSettings().setUseWideViewPort(true);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setPluginState(PluginState.ON);
        webView.getSettings().setSupportZoom(false);
        webView.setScrollBarStyle(WebView.SCROLLBARS_OUTSIDE_OVERLAY);
        webView.loadUrl(url);
    }
}

File: android/app/src/main/kotlin/org/getlantern/lantern/activity/WebViewActivity.kt
package org.getlantern.lantern.activity

import android.os.Bundle
import android.app.ProgressDialog
import android.view.KeyEvent
import android.view.View
import android.webkit.WebSettings.PluginState
import android.webkit.WebView
import android.webkit.WebViewClient
import org.getlantern.lantern.LanternApp
import org.getlantern.lantern.R
import org.getlantern.lantern.model.ProxySettings
import org.getlantern.mobilesdk.Logger

open class WebViewActivity : BaseFragmentActivity() {

    private var dialog: ProgressDialog? = null

    protected var url: String? = null

    lateinit var webView: WebView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.webview)
        webView = findViewById(R.id.webView)
        dialog = ProgressDialog(this)
        url = intent.getStringExtra("url").toString()
        url?.let { openWebview(it) }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (event.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_BACK) {
            if (webView.canGoBack()) webView.goBack() else finish()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    private fun showProgressDialog() {
        dialog?.show()
    }

    private fun hideProgressDialog() {
        dialog?.let { if (it.isShowing) it.dismiss() }
    }

    private fun openWebview(url: String) {
        val proxyAddr: String = LanternApp.getSession().hTTPAddr
        val parts = proxyAddr.split(":").toTypedArray()
        if (parts.size == 2) {
            val proxyHost = parts[0]
            val proxyPort = parts[1].toInt()
            ProxySettings.setProxy(this, webView, proxyHost, proxyPort)
        }
        Logger.d(TAG, "Opening $url in webview")

        // Added to get around an issue where we might be unable
        // to start an intent without a user gesture. This allows
        // the webview to redirect automatically to the payment page
        webView.setWebViewClient(object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                return false
            }
        })
        val settings = webView.getSettings()
        settings.loadWithOverviewMode = true
        settings.useWideViewPort = true
        settings.javaScriptEnabled = true
        settings.pluginState = PluginState.ON
        settings.setSupportZoom(false)
        webView.setScrollBarStyle(WebView.SCROLLBARS_OUTSIDE_OVERLAY)
        webView.loadUrl(url)
    }

    fun closeWebView(view: View) {
        finish()
    }

    companion object {
        private val TAG = WebViewActivity::class.java.simpleName
    }
}
